File: Recrussion/csharp/Fibonacci.cs
using System;
using System.Collections.Generic;

namespace RecursionSolutions;

public static class Fibonacci
{
    private static readonly Dictionary<int, long> Memo = new()
    {
        { 0, 0 },
        { 1, 1 }
    };

    public static long Solve(int n)
    {
        if (n < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(n), "n must be non-negative");
        }

        if (Memo.TryGetValue(n, out var value))
        {
            return value;
        }

        value = Solve(n - 1) + Solve(n - 2);
        Memo[n] = value;
        return value;
    }
}

File: Recrussion/ts/fibonacci.ts
const memo = new Map<number, number>([
  [0, 0],
  [1, 1],
]);

export function fibonacci(n: number): number {
  if (n < 0) {
    throw new Error("n must be non-negative");
  }

  if (memo.has(n)) {
    return memo.get(n)!;
  }

  const value = fibonacci(n - 1) + fibonacci(n - 2);
  memo.set(n, value);
  return value;
}
